File: 242.ValidAnagram/Anagram.java
public class Solution {
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }

        Map<Character, Integer> map = new HashMap<>();
        for (char c : s.toCharArray()) {
            if (!map.containsKey(c)) {
                map.put(c, 0);
            }
            map.put(c, map.get(c)+1);
        }

        for (char c : t.toCharArray()) {
            if (!map.containsKey(c) || map.get(c) == 0) {
                return false;
            }
            else {
                map.put(c, map.get(c)-1);
            }
        }

        return true;
    }
}

File: 242.ValidAnagram/anagram.py
class Solution(object):
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        if len(s) != len(t):
            return False
            
        characters = {}
        for c in s:
            characters[c] = characters.get(c, 0) + 1

        for c in t:
            if c not in characters or characters[c] == 0:
                return False
            else:
                characters[c] -= 1

        return True
